File: 알고리즘 문제풀이/백준/개미전사.py
# dp[i]=max(dp[i-1],d[i-2]+data[i])
def solution(n):
    answer = 0
    data = list(map(int,input().split(" ")))   
    dp=[0]*100
    dp[0],dp[1]=data[0],max(data[0],data[1])
    for i in range(2,n):
        dp[i]=max(dp[i-1],dp[i-2]+data[i])
    #dp = [[0,0] for i in range(n)]
    #dp[0][0],dp[1][1]=data[0],data[1] 
    #for i in range(2,n)
        #dp[i][0]=max(dp[i-1][0],dp[i-2][0]+data[i])
    #for i in range(2,n):
        #dp[i][1]=max(dp[i-1][1],dp[i-2][1]+data[i])
    print(max(dp))
    return answer
  

n = int(input())
solution(n)

File: 알고리즘 문제풀이/Algorithm/dynamic/개미전사.js
/**
1 3 1 5
1 3 3 8 

a[i] 는 i번째 식향 창고까지의 최적의 해
k[i] 는 식량창고에 있는 식량의 양

a[i] = max(a[i-1],a[i-2]+k[i])
*/

const solution = (n, arr) => {
  const dp = new Array(n).fill(0);
  dp[0] = arr[0]; // 0번째 식량창고 까지의 최적의 해
  dp[1] = Math.max(arr[0], arr[1]); // 1번째 식량창고 까지의 최적의 해

  for (let i = 2; i < n; i++) {
    dp[i] = Math.max(dp[i - 1], dp[i - 2] + arr[i]);
  }
  return dp[n - 1];
};

console.log(solution(4, [1, 3, 1, 5]));
